Add explanation for APF debug output (#42075)
kubernetes/website: content/en/docs/tasks/debug/debug-cluster/flow-control.md

Reviewer: I agree with https://github.com/kubernetes/website/pull/42075#issuecomment-1694676318

@mengqiy if you rewrite this to match other pages in the same section in terms of describing steps someone should take, it'll fit better.

Please also set a `weight` for the new page (I would pick a new number higher than any existing weight in the same section).

Author: There are not any steps here. So I have moved it the reference doc as @tengqm suggested.
But I'm not sure what weight I should set. I don't quite understand how it works.

---
title: Troubleshooting APIServer flow control issues
content_type: task
---

<!-- overview -->

API Priority and Fairness controls the behavior of the Kubernetes API server in
an overload situation. You can find more information about it in the

> [concept doc](/docs/concepts/cluster-administration/flow-control.md)

Reviewer: ```suggestion
[API Priority and Fairness](/docs/concepts/cluster-administration/flow-control/)
documentation.
```

## {{% heading "prerequisites" %}}

* Kubernetes cluster is installed
* `kubectl` is configured to communicate with the cluster

> 

Reviewer: I imagine you also need API Priority and Fairness enabled?

Reviewer: ```suggestion
- You must have the `APIPriorityAndFairness`
  [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
  enabled in your cluster (this feature is enabled by default for Kubernetes versions
  1.20 through to {{< skew currentVersion >}}, but you can explicitly disable it).

```

Author: No longer need this part since it's not a troubleshooting doc any more.

<!-- steps -->

## Diagnostics

Every HTTP response from an API server with the priority and fairness feature
enabled has two extra headers: `X-Kubernetes-PF-FlowSchema-UID` and
`X-Kubernetes-PF-PriorityLevel-UID`, noting the flow schema that matched the request
and the priority level to which it was assigned, respectively. The API objects'
names are not included in these headers in case the requesting user does not

> have permission to view them, so when debugging you can use a command like

Reviewer: _optional tweak_
```suggestion
names are not included in these headers (to avoid revealing details in case the
requesting user does not have permission to view them). When debugging, you
can use a command such as:

```shell
kubectl get flowschemas -o custom-columns="uid:{metadata.uid},name:{metadata.name}"
kubectl get prioritylevelconfigurations -o custom-columns="uid:{metadata.uid},name:{metadata.name}"
```

to get a mapping of UIDs to names for both FlowSchemas and

> PriorityLevelConfigurations.

Reviewer: Also consider:
```shell
kubectl get flowschemas -o jsonpath='{range .items[?(.metadata.uid=262bab1a-1c79-11ea-8e23-42010a800016)]}{.metadata.name}{end}'
```
or similar

Author: I'm not a jsonpath expert. But this doesn't seems to be right to output a list of <name, uid> pairs.

## Debug endpoints

When you enable the API Priority and Fairness feature, the `kube-apiserver`
serves the following additional paths at its HTTP(S) ports.

- `/debug/api_priority_and_fairness/dump_priority_levels` - a listing of

>   all the priority levels and the current state of each.  You can fetch like this:

Reviewer: What access do you need in order to fetch that debug info?

Author: Updated

  ```shell
  kubectl get --raw /debug/api_priority_and_fairness/dump_priority_levels
  ```

>   The output is similar to this:

Reviewer: What format is that in? Is it CSV, for example (and defined as such)? If it is, let's say so.

Author: Done

  ```

  Explanation for selected column names:
  - `QueueIndex`: The index of the queue. It will be -1 for priority levels
    without queues.
  - `RequestIndexInQueue`: The index in the queue for a given request. It will
    be -1 for executing requests.
  - `InitialSeats`: The number of seats will be occupied during the initial
    (normal) stage of execution of the request.
  - `FinalSeats`: The number of seats will be occupied during the final stage
    of request execution, accounting for the associated WATCH notifications.
  - `AdditionalLatency`: The extra time taken during the final stage of request
    execution. FinalSeats will be occupied during this time period. It does not
    mean any latency that a user will observe.
  - `StartTime`: The time a request starts to execute. It will be
    0001-01-01T00:00:00Z for queued requests.

## Debug logging

> At `-v=3` or more verbose the server outputs an httplog line for every

Reviewer: ```suggestion
At `-v=3` or more verbosity, the API server outputs an httplog line for every
```

request, and it includes the following attributes.

- `apf_fs`: the name of the flow schema to which the request was classified.
- `apf_pl`: the name of the priority level for that flow schema.
- `apf_iseats`: the number of seats determined for the initial
  (normal) stage of execution of the request.
- `apf_fseats`: the number of seats determined for the final stage of
  execution (accounting for the associated WATCH notifications) of the
  request.
- `apf_additionalLatency`: the duration of the final stage of
  execution of the request.

> 

Reviewer: Where is that log written / how can a cluster operator view it?

Author: In the APIServer log.
I have added it in the doc

At higher levels of verbosity there will be log lines exposing details
of how APF handled the request, primarily for debugging purposes.

## Response headers

APF adds the following two headers to each HTTP response message.

- `X-Kubernetes-PF-FlowSchema-UID` holds the UID of the FlowSchema
  object to which the corresponding request was classified.
- `X-Kubernetes-PF-PriorityLevel-UID` holds the UID of the
  PriorityLevelConfiguration object associated with that FlowSchema.

> 

Reviewer: I would mention whether these appear in audit logs. AIUI they don't, and you can't enable that - have I got this right?

Author: Updated.
Will not appear in audit log.

## {{% heading "whatsnext" %}}

For background information on design details for API priority and fairness, see
the [enhancement proposal](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness).
You can make suggestions and feature requests via [SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery) 

> or the feature's [slack channel](https://kubernetes.slack.com/messages/api-priority-and-fairness).

Reviewer: Skip these details here.
